# Chapter 3: Processing Raw Text

In [132]:
import nltk, re, pprint
from nltk import word_tokenize
from urllib import request

In [133]:
url = 'http://www.gutenberg.org/files/2554/2554-0.txt'
response = request.urlopen(url) 
raw = response.read().decode('utf-8')
print(type(raw), len(raw), raw[:75])

<class 'str'> 1176812 The Project Gutenberg eBook of Crime and Punishment, by Fyodor Dostoevsky


In [134]:
tokens = word_tokenize(raw)
print(type(tokens), len(tokens), tokens[:10])

<class 'list'> 257058 ['\ufeffThe', 'Project', 'Gutenberg', 'eBook', 'of', 'Crime', 'and', 'Punishment', ',', 'by']


In [135]:
text = nltk.Text(tokens)
print(text[1024:1062])
print(text.collocations())

['insight', 'impresses', 'us', 'as', 'wisdom', '...', 'that', 'wisdom', 'of', 'the', 'heart', 'which', 'we', 'seek', 'that', 'we', 'may', 'learn', 'from', 'it', 'how', 'to', 'live', '.', 'All', 'his', 'other', 'gifts', 'came', 'to', 'him', 'from', 'nature', ',', 'this', 'he', 'won', 'for']
Katerina Ivanovna; Pyotr Petrovitch; Pulcheria Alexandrovna; Avdotya
Romanovna; Rodion Romanovitch; Marfa Petrovna; Sofya Semyonovna; old
woman; Project Gutenberg-tm; Porfiry Petrovitch; Amalia Ivanovna;
great deal; young man; Nikodim Fomitch; Project Gutenberg; Ilya
Petrovitch; Andrey Semyonovitch; Hay Market; Dmitri Prokofitch; Good
heavens
None


In [136]:
url1 = 'http://news.bbc.co.uk/2/hi/health/2284783.stm'
html = request.urlopen(url1).read().decode('utf8')
print(type(html), len(html))
print(html[:60])
#print(html)

<class 'str'> 37306
<!doctype html public "-//W3C//DTD HTML 4.0 Transitional//EN


In [137]:
from bs4 import BeautifulSoup 
raw1 = BeautifulSoup(html, 'html.parser').get_text() 
tokens_html = word_tokenize(raw1)
text_html = nltk.Text(tokens_html)
text_html.concordance('gene')
#print(tokens_html)

print(html.find('of the'))

Displaying 7 of 7 matches:
hey say too few people now carry the gene for blondes to last beyond the next 
blonde hair is caused by a recessive gene . In order for a child to have blond
 have blonde hair , it must have the gene on both sides of the family in the g
ere is a disadvantage of having that gene or by chance . They do n't disappear
des would disappear is if having the gene was a disadvantage and I do not thin
er's Polio campaign launched in Iraq Gene defect explains high blood pressure 
er's Polio campaign launched in Iraq Gene defect explains high blood pressure 
24338


In [138]:
hp_1 = open('/Users/maria.onoeva/Desktop/new_folder/HP_1.txt')
raw_HP = hp_1.read()

In [139]:
a = [1, 2, 3, 4, 5, 6, 7, 6, 5, 4, 3, 2, 1]
b = [' ' * 2 * (7 - i) + 'very' * i for i in a]
for line in b:
    print(line)

            very
          veryvery
        veryveryvery
      veryveryveryvery
    veryveryveryveryvery
  veryveryveryveryveryvery
veryveryveryveryveryveryvery
  veryveryveryveryveryvery
    veryveryveryveryvery
      veryveryveryvery
        veryveryvery
          veryvery
            very


In [140]:
my_sentence = 'It\'s the worst case scenario lullaby'
for i in my_sentence.split(' '):
    print(i)
    
for char in my_sentence: 
    print(char, end=' ')

It's
the
worst
case
scenario
lullaby
I t ' s   t h e   w o r s t   c a s e   s c e n a r i o   l u l l a b y 

In [141]:
word_list = [w for w in nltk.corpus.words.words('en') if w.islower()]
# print([w for w in word_list if re.search('ed$', w)])
# print([w for w in word_list if re.search('^..j..t..$', w)])
# print([w for w in word_list if re.search('..j..t..', w)])
# print([w for w in word_list if re.search('^[g-o]+$', w)])
# print([w for w in word_list if re.search('^[g-o]$', w)])
print([w for w in word_list if re.search('^[a-fj]+$', w)])
print([w for w in word_list if re.search('^[a-fj-o]$', w)])
print([w for w in word_list if re.search('^[a-o]$', w)])

['a', 'aa', 'aba', 'abac', 'abaca', 'abaff', 'abb', 'abed', 'acca', 'accede', 'ace', 'ad', 'adad', 'add', 'adda', 'added', 'ade', 'adead', 'ae', 'aface', 'affa', 'ajaja', 'b', 'ba', 'baa', 'baba', 'babe', 'bac', 'bacaba', 'bacca', 'baccae', 'bad', 'bade', 'bae', 'baff', 'bajada', 'be', 'bead', 'beaded', 'bebed', 'bed', 'bedad', 'bedded', 'bedead', 'bedeaf', 'bee', 'beef', 'bejade', 'c', 'ca', 'cab', 'caba', 'cabda', 'cad', 'cade', 'caeca', 'caffa', 'ce', 'cede', 'cee', 'd', 'da', 'dab', 'dabb', 'dabba', 'dace', 'dad', 'dada', 'dade', 'dae', 'daff', 'de', 'dead', 'deaf', 'deb', 'decad', 'decade', 'dee', 'deed', 'deedeed', 'deface', 'e', 'ea', 'ebb', 'ecad', 'edea', 'efface', 'f', 'fa', 'facade', 'face', 'faced', 'fad', 'fade', 'faded', 'fae', 'faff', 'fe', 'fed', 'fee', 'feed', 'j', 'jab', 'jabbed', 'jade', 'jaded', 'jed', 'jeff']
['a', 'b', 'c', 'd', 'e', 'f', 'j', 'k', 'l', 'm', 'n', 'o']
['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o']


In [142]:
print([w for w in word_list if re.search('^..j..t..$', w)])
print([w for w in word_list if re.search(r'^..j..t..$', w)])

['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector', 'unjilted', 'unjolted', 'unjustly']
['abjectly', 'adjuster', 'dejected', 'dejectly', 'injector', 'majestic', 'objectee', 'objector', 'rejecter', 'rejector', 'unjilted', 'unjolted', 'unjustly']


In [143]:
# https://stackoverflow.com/questions/26318287/what-does-r-mean-before-a-regex-pattern
print('\n') # Prints a newline character
print(r'\n') # Escape sequence is not processed
print('\b') # Prints a backspace character
print(r'\b') # Escape sequence is not processed



\n

\b


I guess I got it. So by prefixing with 'r' we kinda say to Python not to read into that but simply to pass a combination to re library, ok!

In [144]:
print([int(n) for n in re.findall(r'[0-9]{2,}', '2009-12-31')])

[2009, 12, 31]


In [145]:
paranoid = """Please could you stop the noise?
I'm tryna get some rest
From all the unborn chicken
Voices in my head"""
pattern = r'^[A-Z]'
pattern1 = r'\?$'
pattern2 = r'es$' # finds nothing because it thinks of the line as the whole string and no line ends with 'es', so in order it to work I have to tokenize it first 

nltk.re_show(pattern, paranoid)

{P}lease could you stop the noise?
{I}'m tryna get some rest
{F}rom all the unborn chicken
{V}oices in my head


In [146]:
from nltk.corpus import brown 
hobbies_learned = nltk.Text(brown.words(categories = ['hobbies', 'learned']))
hobbies_learned.findall(r'<\w*> <and> <other> <\w*s>')
hobbies_learned.findall(r'<as> <\w*> <as> <\w*>')

speed and other activities; water and other liquids; tomb and other
landmarks; Statues and other monuments; pearls and other jewels;
charts and other items; roads and other features; figures and other
objects; military and other areas; demands and other factors;
abstracts and other compilations; iron and other metals
as accurately as possible; as well as the; as faithfully as possible;
as much as what; as neat as a; as simple as you; as well as other; as
well as other; as involved as determining; as well as other; as
important as another; as accurately as possible; as accurate as any;
as much as any; as different as a; as Orphic as that; as coppery as
Delawares; as good as another; as large as small; as well as ease; as
well as their; as well as possible; as straight as possible; as well
as nailed; as smoothly as the; as soon as a; as well as injuries; as
well as many; as well as reason; as well as in; as well as of; as well
as a; as well as summer; as well as providing; as important a

## Exercises 
#### 1. Define a string s = 'colorless'. Write a Python statement that changes this to "colourless" using only the slice and concatenation operations.

In [147]:
s = 'colorless' 
print(s[:4] + 'u' + s[4:])

colourless


#### 2. We can use the slice notation to remove morphological endings on words. For example, 'dogs'[:-1] removes the last character of dogs, leaving dog. Use slice notation to remove the affixes from these words (we've inserted a hyphen to indicate the affix boundary, but omit this from your strings): dish-es, run-ning, nation-ality, un-do, pre-heat.

In [148]:
ex2_words = [w for w in 'dish-es, run-ning, nation-ality, un-do, pre-heat'.split(', ')]
ex2_noaff = [ex2_words[0][:-3], ex2_words[1][:-5], ex2_words[2][:-6], ex2_words[3][-2:], ex2_words[4][-4:]]
print(ex2_noaff)

['dish', 'run', 'nation', 'do', 'heat']


#### 3. We saw how we can generate an IndexError by indexing beyond the end of a string. Is it possible to construct an index that goes too far to the left, before the start of the string? 
I don't think so. I cannot use negative numbers as they remove chars and the lowest possible is then 0 which is the first char.
 
Okay, I see it now. I thought about slicing but not about indexing. So it's possible to raise the error then. 

In [149]:
s = "hello"
print(s[-6:])  # This will print 'hello'
print(s[-100:])  # Still prints 'hello', as slicing is more forgiving
print(s[-1])  # This will print 'o'
print(s[-5])  # This will print 'h'
# Accessing beyond the start (too far to the left)
#print(s[-6])  # This will raise an IndexError

hello
hello
o
h


#### 4. We can specify a "step" size for the slice. The following returns every second character within the slice: monty[6:11:2]. It also works in the reverse direction: monty[10:5:-2] Try these for yourself, then experiment with different step values.

In [150]:
print(paranoid[::2])
print(paranoid[10:5:-2])

Pes ol o tptenie
' rn e oers
rmalteubr hce
ocsi yha
lo 


#### 5. What happens if you ask the interpreter to evaluate monty[::-1]? Explain why this is a reasonable result.
It reverses a string, but why can it be reasonable? Perhaps if some text is reversed, it can reverse it back. Or to print some text so it's readable in mirrors. 

In [151]:
print(paranoid[::-1])
print(paranoid)

daeh ym ni secioV
nekcihc nrobnu eht lla morF
tser emos teg anyrt m'I
?esion eht pots uoy dluoc esaelP
Please could you stop the noise?
I'm tryna get some rest
From all the unborn chicken
Voices in my head


#### 6. Describe the class of strings matched by the following regular expressions. 
1. ```[a-zA-Z]+```: '+' means one or more of previous item, in this case one or more of any letter, case doesn't matter ('cat', 'so') > any word 
2. ```[A-Z][a-z]*```: here case matters, so first it should be an upper one then zero or more of any lower case ('D', 'Dog') > any upper case word 
3. ```p[aeiou]{,2}t```: a sequence should start with 'p' and end with 't', then between there should be at most 2 vowels > too much output, replacing it with ```^p[aeiou]{,2}t$``` ('put', 'poet') 
4. ```\d+(\.\d+)?```: one or more digits and zero or one ('?') of digits after the decimal point
5. ```([^aeiou][aeiou][^aeiou])*```: zero or more instances of a chunk that doesn't start with a vowel then has one vowel and doen't end with the vowel, so its CVC* combinations (??)  -- huh? 
6. ```\w+|[^\w\s]+```: any alphanumeric char OR something that doesn't start with any alphanumeric char and a white space -- anything that is not a space?

In [152]:
nltk.re_show(r'[a-zA-Z]+', paranoid)
nltk.re_show(r'[A-Z][a-z]*', paranoid)
print([w for w in word_list if re.search(r'^p[aeiou]{,2}t$', w)])
print([w for w in word_list if re.search(r'\d+(\.\d+)?', w)])
nltk.re_show(r'([^aeiou][aeiou][^aeiou])*', paranoid)
nltk.re_show(r'\w+|[^\w\s]+', paranoid)

{Please} {could} {you} {stop} {the} {noise}?
{I}'{m} {tryna} {get} {some} {rest}
{From} {all} {the} {unborn} {chicken}
{Voices} {in} {my} {head}
{Please} could you stop the noise?
{I}'m tryna get some rest
{From} all the unborn chicken
{Voices} in my head
['pat', 'paut', 'peat', 'pet', 'piet', 'pit', 'poet', 'poot', 'pot', 'pout', 'put']
[]
{}P{}l{}e{}a{se }{}c{}o{}u{}l{}d{} {}y{}o{}u{} {}s{top}{} {}t{he }{}n{}o{}i{se?}{}
{}I{}'{}m{} {}t{}r{}y{na get}{} {som}{}e{} {res}{}t{}
{}F{rom al}{}l{} {}t{he }{}u{}n{bor}{}n{} {}c{hicken}{}
{}V{}o{}i{ces in}{} {}m{}y{} {}h{}e{}a{}d{}
{Please} {could} {you} {stop} {the} {noise}{?}
{I}{'}{m} {tryna} {get} {some} {rest}
{From} {all} {the} {unborn} {chicken}
{Voices} {in} {my} {head}


### 7. Write regular expressions to match the following classes of strings:

1. A single determiner (assume that a, an, and the are the only determiners).
2. An arithmetic expression using integers, addition, and multiplication, such as 2*3+8.

In [153]:
paranoid_full_raw = open('paranoid_full.txt')
paranoid_full = paranoid_full_raw.read()
nltk.re_show(r'(\ban?\b)|(\bthe\b)', paranoid_full)

Please could you stop {the} noise?
I'm tryna get some rest
From all {the} unborn chicken
Voices in my head
What's that? (I may be paranoid, but not {an} android)
What's that? (I may be paranoid, but not {an} android)
When I am king
You will be first against {the} wall
With your opinion
Which is of no consequence at all
What's that? (I may be paranoid, but no android)
What's that? (I may be paranoid, but no android)
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la-la-la-la
La, la
Ambition makes you look pretty ugly
Kicking, squealing, Gucci little piggy
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la
You don't remember, you don't remember
Why don't you remember my name?
Off with his head, man, off with his head, man
Why don't you remember my name?
I guess he does
Ah, oh, oh, oh
Oh, oh, oh, oh
Oh, oh
Oh, oh
Rain down, rain down
Come on, rain down on me
From {a} great height
From {a} great height, height
Rain down, rain down
Come on, rain down on me
From {a} great height
From {

In [154]:
arithmetics_7b = "2+2+2+2+2, 2+2, 5*4, 73-74+73-74, 34/2, 2*3+8, banana, apple, DNA, 451Fahrenheit" # {451}Fahrenheit is picked in both cases

nltk.re_show(r'-?\d+(\s*[\+\*]\s*-?\d+)*', arithmetics_7b) # ChatGPT
nltk.re_show(r'([^\sa-zA-Z,]*[\+\*]*)*[^\s,a-zA-Z]', arithmetics_7b) # my version 

{2+2+2+2+2}, {2+2}, {5*4}, {73}{-74+73}{-74}, {34}/{2}, {2*3+8}, banana, apple, DNA, {451}Fahrenheit
{2+2+2+2+2}, {2+2}, {5*4}, {73-74+73-74}, {34/2}, {2*3+8}, banana, apple, DNA, {451}Fahrenheit


#### 8. Write a utility function that takes a URL as its argument, and returns the contents of the URL, with all HTML markup removed. Use from urllib import request and then request.urlopen('http://nltk.org/').read().decode('utf8') to access the contents of the URL.

In [155]:
def url_opener(link):
    url = link 
    response = request.urlopen(url) 
    raw = response.read().decode('utf8')
    ready_url = BeautifulSoup(raw, 'html.parser').get_text()
    print(ready_url) 
    
url_opener('http://nltk.org/')
url_opener('https://mariaonoeva.github.io/Biennial2024/')







NLTK :: Natural Language Toolkit















NLTK



Documentation














NLTK Documentation

API Reference
Example Usage
Module Index
Wiki
FAQ
Open Issues
NLTK on GitHub

Installation

Installing NLTK
Installing NLTK Data

More

Release Notes
Contributing to NLTK
NLTK Team






Natural Language Toolkit¶
NLTK is a leading platform for building Python programs to work with human language data.
It provides easy-to-use interfaces to over 50 corpora and lexical
resources such as WordNet,
along with a suite of text processing libraries for classification, tokenization, stemming, tagging, parsing, and semantic reasoning,
wrappers for industrial-strength NLP libraries,
and an active discussion forum.
Thanks to a hands-on guide introducing programming fundamentals alongside topics in computational linguistics, plus comprehensive API documentation,
NLTK is suitable for linguists, engineers, students, educators, researchers, and industry users alike.
NLTK is available for Window

#### 9.  Save some text into a file corpus.txt. Define a function load(f) that reads from the file named in its sole argument, and returns a string containing the text of the file.

1. Use nltk.regexp_tokenize() to create a tokenizer that tokenizes the various kinds of punctuation in this text. Use one multi-line regular expression, with inline comments, using the verbose flag (?x).

In [156]:
def load(my_text_raw): 
    my_text = open(my_text_raw)
    my_text_read = my_text.read() 
    return my_text_read

def punct_counter(my_text):
    counter_list = []
    punct = nltk.regexp_tokenize(load(my_text), r'[^\w\s\d]')
    for i in punct: 
        counter_list.append(punct.count(i))
    ready_dict = dict(zip(punct, counter_list))
    sorted_dict = dict(sorted(ready_dict.items(), key=lambda item: item[1], reverse=True))
    return sorted_dict.items()

punct_counter('paranoid_full.txt')
eng_HP = punct_counter('/Users/maria.onoeva/Desktop/new_folder/HP_1.txt')
ru_HP = punct_counter('/Users/maria.onoeva/Desktop/new_folder/HP_1ru.txt')

In [157]:
from tabulate import tabulate 
headers = ['Punctuation', 'Count']

print(tabulate(eng_HP, headers=headers))
print(tabulate(ru_HP, headers=headers))

# I want them side by side 


Punctuation      Count
-------------  -------
.                 6136
,                 5658
"                 4758
'                 3141
-                 1990
?                  754
!                  474
;                  135
:                   69
)                   33
(                   30
*                    2
~                    1
\                    1
Punctuation      Count
-------------  -------
,                10190
.                 6355
—                 4598
-                 1038
…                  805
?                  761
!                  620
«                  222
»                  221
:                  140
*                  117
(                   13
)                   13
;                    3
№                    1
“                    1
”                    1


#### Harry Potter 1 punctuation: Eng and Ru 
##### Eng: 
- why are there 30 of '(' and 33 of ')'? 
- questions: 754
##### Ru: 
- same for Ru for kavychki, but skobki are even
- so many commas??? 
- questions: 761 > 7 more > pretty consistent, I can imagine contexts where several questions mark used to express anger or surprise 

In [158]:
# ChatGPT to display the tables side by side
# Generate table strings
table1 = tabulate(eng_HP, headers=['Punctuation ENG', 'Count'], tablefmt='plain')
table2 = tabulate(ru_HP, headers=['Punctuation RU', 'Count'], tablefmt='plain')

# Split tables into lines
lines1 = table1.split('\n')
lines2 = table2.split('\n')

# Determine the max width of the first table
max_width = max(len(line) for line in lines1)

# Determine the longer table for iteration
longest_table_length = max(len(lines1), len(lines2))

# Print tables side by side, handling uneven number of rows
for i in range(longest_table_length):
    line1 = lines1[i] if i < len(lines1) else ""
    line2 = lines2[i] if i < len(lines2) else ""
    print(f"{line1.ljust(max_width)}    {line2}")

Punctuation ENG      Count    Punctuation RU      Count
.                     6136    ,                   10190
,                     5658    .                    6355
"                     4758    —                    4598
'                     3141    -                    1038
-                     1990    …                     805
?                      754    ?                     761
!                      474    !                     620
;                      135    «                     222
:                       69    »                     221
)                       33    :                     140
(                       30    *                     117
*                        2    (                      13
~                        1    )                      13
\                        1    ;                       3
                              №                       1
                              “                       1
                              ”                 

#### 9. Save some text into a file corpus.txt. Define a function load(f) that reads from the file named in its sole argument, and returns a string containing the text of the file.

2. Use nltk.regexp_tokenize() to create a tokenizer that tokenizes the following kinds of expression: monetary amounts; dates; names of people and organizations.

In [159]:
# monetary amounts? 
monetary_amounts = "200$, 363524$, 587$, 65€, 8746CZK, 8₽, 9382₽, 87362513537CZK, Grand Theft Auto V, Corpus paper, April, Thursday, 76F, 17C"
money_pattern = r'\d+([\$€₽]|CZK)+' 

nltk.re_show(money_pattern, monetary_amounts)

{200$}, {363524$}, {587$}, {65€}, {8746CZK}, {8₽}, {9382₽}, {87362513537CZK}, Grand Theft Auto V, Corpus paper, April, Thursday, 76F, 17C


In [160]:
# dates
dates_more = '2024-04-04, 2024-04-05, Your Top Songs 2023, 2024-04-06, pinguin, 2024-04-07, 2024-04-08, Ice-cream, 2024-04-09, 2024-04-10, Sněžka'
dates_pattern = r'\d{4}-\d{2}-\d{2}'
nltk.re_show(dates_pattern, dates_more)

# names of people and organizations
names_pattern = r'[A-Z]+'
nltk.re_show(names_pattern, dates_more)

{2024-04-04}, {2024-04-05}, Your Top Songs 2023, {2024-04-06}, pinguin, {2024-04-07}, {2024-04-08}, Ice-cream, {2024-04-09}, {2024-04-10}, Sněžka
2024-04-04, 2024-04-05, {Y}our {T}op {S}ongs 2023, 2024-04-06, pinguin, 2024-04-07, 2024-04-08, {I}ce-cream, 2024-04-09, 2024-04-10, {S}něžka


#### 10. Rewrite the following loop as a list comprehension:

In [161]:
sent = ['The', 'dog', 'gave', 'John', 'the', 'newspaper']
result = []
for word in sent:
    word_len = (word, len(word))
    result.append(word_len)

print(result)

[('The', 3), ('dog', 3), ('gave', 4), ('John', 4), ('the', 3), ('newspaper', 9)]


In [162]:
result_1 = [(word, len(word)) for word in sent] 
print(result_1)

[('The', 3), ('dog', 3), ('gave', 4), ('John', 4), ('the', 3), ('newspaper', 9)]


#### 11. Define a string raw containing a sentence of your own choosing. Now, split raw on some character other than space, such as 's'.

In [163]:
born_this_way = """Don't be a drag, just be a queen
Whether you're broke or evergreen
You're Black, white, beige, chola descent
You're Lebanese, you're Orient'
Whether life's disabilities left you outcast, bullied, or teased
Rejoice and love yourself today
'Cause, baby, you were born this way""" 

print(born_this_way.split('\n'))

["Don't be a drag, just be a queen", "Whether you're broke or evergreen", "You're Black, white, beige, chola descent", "You're Lebanese, you're Orient'", "Whether life's disabilities left you outcast, bullied, or teased", 'Rejoice and love yourself today', "'Cause, baby, you were born this way"]


#### 12. Write a for loop to print out the characters of a string, one per line.

In [164]:
dna = "Same DNA, but born this way"
for char in dna: 
    print(char)

S
a
m
e
 
D
N
A
,
 
b
u
t
 
b
o
r
n
 
t
h
i
s
 
w
a
y


#### 13. What is the difference between calling split on a string with no argument or with ' ' as the argument, e.g. sent.split() versus sent.split(' ')? What happens when the string being split contains tab characters, consecutive space characters, or a sequence of tabs and spaces? (In IDLE you will need to use '\t' to enter a tab character.)

In [165]:
print(dna.split()) # space split
print(dna.split(' ')) # same as the empty argument 
print(dna.split('\t')) # no change because there is no tab

['Same', 'DNA,', 'but', 'born', 'this', 'way']
['Same', 'DNA,', 'but', 'born', 'this', 'way']
['Same DNA, but born this way']


#### 14. Create a variable words containing a list of words. Experiment with words.sort() and sorted(words). What is the difference?
https://www.geeksforgeeks.org/python-difference-between-sorted-and-sort/ -- cool! 

In [166]:
dna_split = dna.split()
print(dna_split.sort()) # none? why? > as per documentation returns None and changes the initial list 
print(sorted(dna_split)) # only returns the sorted list and makes no changes 

None
['DNA,', 'Same', 'born', 'but', 'this', 'way']


#### 15.  Explore the difference between strings and integers by typing the following at a Python prompt: "3" * 7 and 3 * 7. Try converting between strings and integers using int("3") and str(3).

In [167]:
print("3"*7)
print(3*7)
print(int("3")*7)
print(str(3)*7)

3333333
21
21
3333333


#### 16. Use a text editor to create a file called prog.py containing the single line monty = 'Monty Python'. Next, start up a new session with the Python interpreter, and enter the expression monty at the prompt. You will get an error from the interpreter. Now, try the following (note that you have to leave off the .py part of the filename).  This time, Python should return with a value. You can also try import prog, in which case Python should be able to evaluate the expression prog.monty at the prompt.

In [168]:
from prog import monty
monty

'Monty Python'

#### 17. What happens when the formatting strings %6s and %-6s are used to display strings that are longer than six characters?
I don't get this... 

In [169]:
print("Number is : %6s" % 'thisisaword')
print("Number is : %-6s" % 'thisisaword')
print("Number is : %6s" % 'aword')
print("Number is : %-6s" % 'aword')

Number is : thisisaword
Number is : thisisaword
Number is :  aword
Number is : aword 


#### 18. Read in some text from a corpus, tokenize it, and print the list of all wh-word types that occur. (wh-words in English are used in questions, relative clauses and exclamations: who, which, what, and so on.) Print them in order. Are any words duplicated in this list, because of the presence of case distinctions or punctuation?
Woooow, I did it already in Chapter 2 but in a different way! 
- Well, I have some non-wh-words (whisperers, whirled, wheel, etc.), should I remove them? How to that without a wh-words list as in Chapter 2? 
- At the beginning I wanted to sum all duplicates, but now I sort of like that I have upper and lower case as separate lines because it shows different uses (interrogative and relative wh-words) 
- Updated wh_counter function, now it gives a table right away
- idk i'll think about it later, the list way from Chapter 2 was easier in a way 

In [170]:
import pandas as pd 
def wh_counter(my_text):
    counter_list = []
    wh_raw = nltk.regexp_tokenize(load(my_text), r'^[Ww]h[a-z]+')
    for i in wh_raw: 
        counter_list.append(wh_raw.count(i))
    ready_dict = dict(zip(wh_raw, counter_list))
    sorted_dict = dict(sorted(ready_dict.items(), key=lambda item: item[1], reverse=True))
    output_wh_counter = pd.DataFrame(sorted_dict.items())
    headers_wh = ['Wh-word', 'Count']
    print(tabulate(output_wh_counter, headers=headers_wh))

wh_counter('/Users/maria.onoeva/Desktop/new_folder/HP_1.txt')

    Wh-word       Count
--  ----------  -------
 0  When             20
 1  when             13
 2  what             13
 3  who              10
 4  where             8
 5  What              8
 6  which             6
 7  while             6
 8  Where             6
 9  whether           4
10  While             3
11  whatever          3
12  white             2
13  wheel             1
14  whisperers        1
15  whistled          1
16  Who               1
17  whole             1
18  Whether           1
19  Whispers          1
20  Wheeling          1
21  whirled           1
22  Which             1
23  why               1
24  whirl             1
25  Whatever          1


#### 19. Create a file consisting of words and (made up) frequencies, where each line consists of a word, the space character, and a positive integer, e.g. fuzzy 53. Read the file into a Python list using open(filename).readlines(). Next, break each line into its two fields using split(), and convert the number into an integer using int(). The result should be a list of the form: [['fuzzy', 53], ...]. 

In [171]:
words_19 = load('chap3_ex19.txt')
ex19 = [item.split() for item in words_19.split('\n')]
ex19_1 = [[item[0], int(item[1])] for item in ex19]

print(ex19_1)

[['current', 56], ['file', 32], ['heat', 85], ['moment', 12], ['by', 47], ['my', 36], ['side', 93]]


#### 20. Write code to access a favorite webpage and extract some text from it. For example, access a weather site and extract the forecast top temperature for your town or city today. 

In [172]:
def url_opener_Return(link):
    url = link 
    response = request.urlopen(url) 
    raw = response.read().decode('utf8')
    ready_url = BeautifulSoup(raw, 'html.parser').get_text()
    return ready_url
    
url_opener_Return('https://www.bbc.com/weather/3067696')[1375:1596]

'PragueDay by day forecastLast updated today at 14:00Today, Light rain and a moderate breezeLight RainLight Rain, High7° 45°Low-2° 28°, Wind speed15 mph23 km/h N15 mph23 km/hNortherlyLight rain and a moderate breezeSunday\xa0'

#### 21. Write a function unknown() that takes a URL as its argument, and returns a list of unknown words that occur on that webpage. In order to do this, extract all substrings consisting of lowercase letters (using re.findall()) and remove any items from this set that occur in the Words Corpus (nltk.corpus.words). Try to categorize these words manually and discuss your findings.

In [173]:
def unknown(link):
    link_text = url_opener_Return(link)
    lower_all_words = [w.lower() for w in nltk.corpus.words.words()]
    list_words = [w.lower() for w in word_tokenize(link_text) if w.isalpha()]
    final_list = []
    for word in list_words:
        if word not in lower_all_words:
            final_list.append(word)    
    print(set(final_list))

In [174]:
unknown('https://www.cyberpunk.net/us/en/')

{'things', 'logo', 'rights', 'agreementfan', 'gamespot', 'breakdownmore', 'trademarks', 'rpg', 'futureimmerse', 'years', 'edgerunners', 'reboot', 'morewatch', 'libertyshowsedgerunnerscommunityforumsdiscordcosplay', 'addressenter', 'projekt', 'declarationredmodenglishenglishрусскийdeutschpolskiportuguês', 'countries', 'rewardsmerchandisemediasupportcd', 'policycareersuser', 'cyberpunk', 'games', 'trailerlearn', 'netrunner', 'offers', 'onterms', 'policysubmitfind', 'email', 'routes', 'br', 'françaisespañolespañol', 'guidesbuild', 'italiano日本語한국어简体中文繁體中文العربيةbuy', 'morelearn', 'stories', 'plannergame', 'cd', 'moregamescyberpunk', 'bookletmoreupdate', 'announcements', 'guidelinescookie', 'storyline', 'redenenglishрусскийdeutschpolskiportuguês', 'gx', 'experiencelearn', 'eurogamerget', 'mx'}


#### 22. Examine the results of processing the URL http://news.bbc.co.uk/ using the regular expressions suggested above. You will see that there is still a fair amount of non-textual data there, particularly Javascript commands. You may also find that sentence breaks have not been properly preserved. Define further regular expressions that improve the extraction of text from this web page.

In [175]:
bbc_web = url_opener_Return('http://news.bbc.co.uk/')
nltk.re_show(r'[A-Z][a-z]*', bbc_web)

{Home} - {B}{B}{C} {News}{B}{B}{C} {Homepage}{Skip} to content{Accessibility} {Help}{Your} account{Home}{News}{Sport}{Earth}{Reel}{Worklife}{Travel}{More} menu{More} menu{Search} {B}{B}{C}{Home}{News}{Sport}{Earth}{Reel}{Worklife}{Travel}{Culture}{Future}{Music}{T}{V}{Weather}{Sounds}{Close} menu{B}{B}{C} {News}{Menu}{Home}{Israel}-{Gaza} war{War} in {Ukraine}{India} {Election} 2024{Climate}{Video}{World}{U}{K}{Business}{Tech}{More}{Science}{Entertainment} & {Arts}{Health}{World} {News} {T}{V}{In} {Pictures}{B}{B}{C} {Verify}{Newsbeat}{Sign} {In} {Banner}{Discover} your {B}{B}{C}{Sign} in or create an account to watch, listen and join in{Sign} inor{Register}{Close} sign in banner{B}{B}{C} {News}{An} audible sigh of relief in the {Middle} {East}{The} latest round in the longstanding, dangerous rivalry between {Israel} and {Iran} seems over for now, {Lyse} {Doucet} writes.{Attribution}{Middle} {East}{Posted}10 hours ago10h{Live}. {Iran} downplays presumed {Israeli} attack but vows respon

#### 23. Are you able to write a regular expression to tokenize text in such a way that the word don't is tokenized into do and n't? Explain why this regular expression won't work: «n't|\w+».
It works, but it also matches all dos as in 'down' etc. Can I fix it? -- yes! using lookahead '?=n\'t'

In [176]:
ex22 = r'n\'t|\w+' 
ex22_me = r'don\'t'
ex22_me1 = r'n\'t|do(?=n\'t)'
nltk.re_show(ex22_me1, paranoid_full)

Please could you stop the noise?
I'm tryna get some rest
From all the unborn chicken
Voices in my head
What's that? (I may be paranoid, but not an android)
What's that? (I may be paranoid, but not an android)
When I am king
You will be first against the wall
With your opinion
Which is of no consequence at all
What's that? (I may be paranoid, but no android)
What's that? (I may be paranoid, but no android)
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la-la-la-la
La, la
Ambition makes you look pretty ugly
Kicking, squealing, Gucci little piggy
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la-la-la-la
La-la-la
You {do}{n't} remember, you {do}{n't} remember
Why {do}{n't} you remember my name?
Off with his head, man, off with his head, man
Why {do}{n't} you remember my name?
I guess he does
Ah, oh, oh, oh
Oh, oh, oh, oh
Oh, oh
Oh, oh
Rain down, rain down
Come on, rain down on me
From a great height
From a great height, height
Rain down, rain down
Come on, rain down on me
From a great height
From a

#### 24. Try to write code to convert text into hAck3r, using regular expressions and substitution, where e → 3, i → 1, o → 0, l → |, s → 5, . → 5w33t!, ate → 8. Normalize the text to lowercase before converting it. Add more substitutions of your own. Now try to map s to two different values: $ for word-initial s, and 5 for word-internal s.

In [177]:
paranoid_full_ex24 = paranoid_full.lower() 

# idk how to do it in a more sophisticated way
paranoid_full_ex24 = re.sub(r'e', '3', paranoid_full_ex24)
paranoid_full_ex24 = re.sub(r'i', '1', paranoid_full_ex24)
paranoid_full_ex24 = re.sub(r'o', '1', paranoid_full_ex24)   
paranoid_full_ex24 = re.sub(r'l', '|', paranoid_full_ex24)
paranoid_full_ex24 = re.sub(r'\.', '5w33t!', paranoid_full_ex24)
paranoid_full_ex24 = re.sub(r'ate', '8', paranoid_full_ex24)


In [178]:
paranoid_full_ex24 = paranoid_full.lower() 
paranoid_full_ex24 = re.sub(r'\ss', ' $', paranoid_full_ex24)
paranoid_full_ex24 = re.sub(r'(?<!\b)s', '5', paranoid_full_ex24)
print(paranoid_full_ex24)

plea5e could you $top the noi5e?
i'm tryna get $ome re5t
from all the unborn chicken
voice5 in my head
what's that? (i may be paranoid, but not an android)
what's that? (i may be paranoid, but not an android)
when i am king
you will be fir5t again5t the wall
with your opinion
which i5 of no con5equence at all
what's that? (i may be paranoid, but no android)
what's that? (i may be paranoid, but no android)
la-la-la-la-la-la
la-la-la-la-la-la
la-la-la-la-la-la
la, la
ambition make5 you look pretty ugly
kicking, $quealing, gucci little piggy
la-la-la-la-la-la
la-la-la-la-la-la
la-la-la-la-la-la
la-la-la
you don't remember, you don't remember
why don't you remember my name?
off with hi5 head, man, off with hi5 head, man
why don't you remember my name?
i gue55 he doe5
ah, oh, oh, oh
oh, oh, oh, oh
oh, oh
oh, oh
rain down, rain down
come on, rain down on me
from a great height
from a great height, height
rain down, rain down
come on, rain down on me
from a great height
from a great height, h

In [179]:
# from ChatGPT > I don't know '|'.join, re.escape and re.compile and don't remember how lambda works :( 
# Define the replacements
replacements = {
    'e': '3',
    'i': '1',
    'o': '1',
    'l': '|',
    's': '5',
    '.': '5w33t!',
    'ate': '8'
}

# Define a function to perform the replacements
def replace_chars(text, replacements):
    pattern = re.compile('|'.join(re.escape(key) for key in replacements.keys()))
    return pattern.sub(lambda x: replacements[x.group()], text)

# Apply the replacements
paranoid_full_ex24_gpt = replace_chars(paranoid_full_ex24, replacements)
print(paranoid_full_ex24_gpt)

p|3a53 c1u|d y1u $t1p th3 n1153?
1'm tryna g3t $1m3 r35t
fr1m a|| th3 unb1rn ch1ck3n
v11c35 1n my h3ad
what'5 that? (1 may b3 paran11d, but n1t an andr11d)
what'5 that? (1 may b3 paran11d, but n1t an andr11d)
wh3n 1 am k1ng
y1u w1|| b3 f1r5t aga1n5t th3 wa||
w1th y1ur 1p1n11n
wh1ch 15 1f n1 c1n53qu3nc3 at a||
what'5 that? (1 may b3 paran11d, but n1 andr11d)
what'5 that? (1 may b3 paran11d, but n1 andr11d)
|a-|a-|a-|a-|a-|a
|a-|a-|a-|a-|a-|a
|a-|a-|a-|a-|a-|a
|a, |a
amb1t11n mak35 y1u |11k pr3tty ug|y
k1ck1ng, $qu3a|1ng, gucc1 |1tt|3 p1ggy
|a-|a-|a-|a-|a-|a
|a-|a-|a-|a-|a-|a
|a-|a-|a-|a-|a-|a
|a-|a-|a
y1u d1n't r3m3mb3r, y1u d1n't r3m3mb3r
why d1n't y1u r3m3mb3r my nam3?
1ff w1th h15 h3ad, man, 1ff w1th h15 h3ad, man
why d1n't y1u r3m3mb3r my nam3?
1 gu355 h3 d135
ah, 1h, 1h, 1h
1h, 1h, 1h, 1h
1h, 1h
1h, 1h
ra1n d1wn, ra1n d1wn
c1m3 1n, ra1n d1wn 1n m3
fr1m a gr3at h31ght
fr1m a gr3at h31ght, h31ght
ra1n d1wn, ra1n d1wn
c1m3 1n, ra1n d1wn 1n m3
fr1m a gr3at h31ght
fr1m a gr3at h31ght, h